sharding/colocation.rst

```diff
@@ -0,0 +1,159 @@
+
+.. code-block:: postgresql
+
+  -- create Citus tables, naively use event_id and page_id as distribution columns
+  SELECT create_distributed_table('event', 'event_id');
+  SELECT create_distributed_table('page', 'page_id');
+
+Given that the data is dispersed across different workers, we cannot simply perform a join as we would on a single PostgreSQL node. Instead, we will need to issue two queries:
+
+Across all shards of the page table (Q1):
+
+.. code-block:: postgresql
+
+  SELECT page_id FROM page WHERE path LIKE '/blog%' AND tenant_id = 6;
+
+Across all shards of the event table (Q2):
+
+.. code-block:: postgresql
+
+  SELECT page_id, count(*) AS count
+  FROM events
+  WHERE page_id IN (ARRAY[...page IDs from first query...]) AND 
+        tenant_id = 6 AND 
+        payload->>'time' >= now() - interval '1 week'
+  GROUP BY page_id ORDER BY count DESC LIMIT 10;
+
+
+Afterwards, the results from the two steps need to be combined by the application.
+
+The data required to answer the query is scattered across the shards on the different nodes and each of those shards will need to be queried:
+
+.. image:: ../images/colocation-inefficient-queries.png
+
+In this case the data distribution creates substantial drawbacks:
+
+* Overhead from querying each shard, running multiple queries
+* Overhead of Q1 returning many rows to the client
+* Q2 becoming very large
+* The need to write queries in multiple steps, combine results, requires changes in the application
+
+A potential upside of the relevant data being dispersed is that the queries can be parallelised, which Citus will do. However, this is only beneficial if the amount of work that the query does is substantially greater than the overhead of querying many shards. It’s generally better to avoid doing such heavy lifting directly from the application, for example by `pre-aggregating <insert_into_select>`_ the data.
+
+Distributing tables by tenant
+-----------------------------
+
+Looking at our query again, we can see that all the rows that the query needs have one dimension in common: :code:`tenant_id`. The dashboard will only ever query for a tenant’s own data. That means that if data for the same tenant are always co-located on a single PostgreSQL node, our original query could be answered in a single step by that node by performing a join on :code:`tenant_id` and :code:`page_id`.
+
+In Citus, rows with the same distribution column value are guaranteed to be on the same node. Each shard in a distributed table effectively has a set of co-located shards from other distributed tables that contain the same distribution column values (data for the same tenant). Starting over, we can create our tables with :code:`tenant_id` as the distribution column.
+
+.. code-block:: postgresql
+
+  -- create Citus tables, co-locate them by using a common distribution column
```

Comment:
Could we update to use the explicit co-location API?

So: (a) Update comment to something like "create Citus tables, co-locate them by using a common distribution column and the co-locate-with parameter to ensure physical co-located placement"
and (b), change co-block to:
``` SQL
SELECT create_distributed_table('event', 'tenant_id');
SELECT create_distributed_table('page', 'tenant_id', colocate_with => 'event');
```